src/realm/array_integer.cpp

```diff
@@ -29,6 +30,112 @@ Mixed ArrayInteger::get_any(size_t ndx) const
+            i += 1;
+        }
+
+        return true;
+    }
+    return false;
+}
+
+bool ArrayInteger::try_to_compress_array(std::vector<int64_t>& values, std::vector<size_t>& indices)
+{
+    std::set<int64_t> s;
+    const auto sz = size();
+    indices.reserve(sz);
+
+    for (size_t i = 0; i < sz; ++i) {
+        auto item = get(i);
+        indices.push_back(item);
+        s.insert(item);
+    }
+
+    values.reserve(s.size());
+    for (const auto& v : s) {
+        values.push_back(v);
+    }
+
+    for (auto& v : indices) {
+        auto pos = std::lower_bound(values.begin(), values.end(), v);
+        v = std::distance(values.begin(), pos);
+    }
+
+    auto max = std::max_element(values.begin(), values.end());
+    auto compressed_value_width = bit_width(*max);
+    auto compressed_index_width = bit_width(indices.size() - 1);
+    auto compressed_values_size = compressed_value_width * values.size();
+    auto compressed_indices_size = compressed_index_width * indices.size();
+    // auto total_compressed_size = compressed_values_size + compressed_indices_size;
+    // comment this check for now in order to force compression (needed for testing)
+    // if(total_compressed_size < bit_width(*max)*size() )
+    {
+        m_compressed_value_width = compressed_value_width;
+        m_compressed_index_width = compressed_index_width;
+        m_compressed_values_size = compressed_values_size;
+        m_compressed_indices_size = compressed_indices_size;
+        m_compressed_array = m_alloc.alloc(m_compressed_values_size + m_compressed_indices_size);
+        return true;
+    }
+    return false;
+}
+
+bool ArrayInteger::decompress()
+{
+    if (m_is_compressed) {
+        // decompress
+
+        // this is needed right now because I am keeping the elements underneath in the array
+        // since I need to verify that all the APIs for finding and in general querying are
+        // working
+        clear();
+        const auto sz = m_compressed_indices_size / m_compressed_index_width;
+        const auto offset = m_compressed_values_size;
+        for (size_t i = 0; i < sz; ++i) {
+            auto index = get_direct(m_compressed_array.get_addr(), m_compressed_index_width, offset + i);
+            auto value = get_direct(m_compressed_array.get_addr(), m_compressed_value_width, index);
+            // auto index = read_compressed_value(m_compressed_index_width, m_compressed_array, offset + i);
+            // auto value = read_compressed_value(m_compressed_value_width, m_compressed_array, index);
+            Array::insert(i, value);
+        }
+        m_is_compressed = false;
+
+        return true;
+    }
+    return false;
+}
+
+int64_t ArrayInteger::get_compressed_value(size_t ndx) const
+{
+    const auto sz = m_compressed_indices_size / m_compressed_index_width;
+    const auto offset = m_compressed_values_size;
+    for (size_t i = 0; i < sz; ++i) {
```

Comment:
this is backwards. You shouldn't need a loop here. Just a direct loop up of the index, then using that index for a direct lookup of the value.